### PR TITLE
Add CVMFS_S3_X_AMZ_ACL

### PR DESCRIPTION
**Pull request by mharvey-jt**
Permitted values:
```    "private",
    "public-read",
    "public-write",
    "authenticated-read",
    "aws-exec-read",
    "bucket-owner-read",
    "bucket-owner-full-control",
    ""
```
default is `public-read`

### CODE REVIEW

**Comment by HereThereBeDragons** [comment]
Issue #2970

**Comment by jblomer** [comment]
@cernvm-bot unittest

**Comment by jblomer** [comment]
@cernvm-bot cloudtest

**Comment by jblomer** [comment]
@cernvm-bot unittest

**Comment by jblomer** [comment]
@cernvm-bot tidy

**Comment by jblomer** [comment]
@cernvm-bot cpplint

**Comment by HereThereBeDragons** [comment]
@jblomer there are some typos that prevent building this PR. i fixed it and open a PR on mharvey-jt branch

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

File: cvmfs/upload_s3.h [hunk 1]
```diff
@@ -136,6 +136,22 @@ class S3Uploader : public AbstractUploader {
   const std::string temporary_path_;
   mutable atomic_int32 io_errors_;
   pthread_t thread_collect_results_;
+
+  std::string x_amz_acl_;
+
+  /**
+   *  * Allowed values of x-amz-acl according to S3 API
```
Comment (jblomer) [inline cvmfs/upload_s3.h]:
```suggestion
   * Allowed values of x-amz-acl according to S3 API
```

File: cvmfs/upload_s3.h [hunk 2]
```diff
@@ -136,6 +136,22 @@ class S3Uploader : public AbstractUploader {
...
+   *  */
```
Comment (jblomer) [inline cvmfs/upload_s3.h]:
```suggestion
   **/
```

File: cvmfs/upload_s3.h [hunk 3]
```diff
@@ -136,6 +136,22 @@ class S3Uploader : public AbstractUploader {
...
+  std::string const  x_amz_acl_allowed_values_[8] = {
```
Comment (jblomer) [inline cvmfs/upload_s3.h]: Let's make this `static const char *`.

File: cvmfs/s3fanout.cc [hunk 4]
```diff
@@ -604,7 +612,7 @@ bool S3FanoutManager::MkV4Authz(const JobInfo &info, vector<string> *headers)
   string signing_key = GetAwsV4SigningKey(date);
   string signature = shash::Hmac256(signing_key, string_to_sign);
 
-  headers->push_back("X-Amz-Acl: public-read");
+  headers->push_back("X-Amz-Acl: "+config_.x_amz_acl);
```
Comment (jblomer) [inline cvmfs/s3fanout.cc]:
```suggestion
  headers->push_back("X-Amz-Acl: " + config_.x_amz_acl);
```

**Review by jblomer — CHANGES_REQUESTED** [review]
In principal looks good to me!  One nit on the array of allowed values.  We could use test case 684 to test that an unknown value will be rejected. But there isn't that much value in the test.  I guess the per-bucket policy can only be tested against GCS? In this case, let's not bother for the moment.

**Comment by mharvey-jt** [comment]
the most straightforward way to test this would be to have the test harness apache log transaction headers and then confirm the expected ones are present. Is there precedent for that elsewhere in the test suite?

**Comment by jblomer** [comment]
@cernvm-bot cloudtest

**Pull request merged**